truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1360,153 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
         }
     }
 
+    @Primitive(name = "string_scrub")
+    @ImportStatic(StringGuards.class)
+    public abstract static class ScrubNode extends PrimitiveArrayArgumentsNode {
+
+        @Child private YieldNode yieldNode = new YieldNode();
+        @Child private RopeNodes.MakeConcatNode makeConcatNode = RopeNodesFactory.MakeConcatNodeGen.create(null, null, null);
+
+        @Specialization(guards = { "!is7Bit(string)", "!isValidCodeRange(string)", "isAsciiCompatible(string)" })
+        public DynamicObject scrubAsciiCompat(VirtualFrame frame, DynamicObject string, DynamicObject block) {
+            final Rope rope = rope(string);
+            final Encoding enc = rope.getEncoding();
+            Rope buf = RopeConstants.EMPTY_ASCII_8BIT_ROPE;
+
+            final byte[] pBytes = rope.getBytes();
+            final int e = pBytes.length;
+
+            int p = 0;
+            int p1 = 0;
+
+            p = StringSupport.searchNonAscii(pBytes, p, e);
+            if (p == -1) {
+                p = e;
+            }
+            while (p < e) {
+                int ret = enc.length(pBytes, p, e);
+                if (MBCLEN_NEEDMORE_P(ret)) {
+                    break;
+                } else if (MBCLEN_CHARFOUND_P(ret)) {
+                    p += MBCLEN_CHARFOUND_LEN(ret);
+                } else if (MBCLEN_INVALID_P(ret)) {
+                    // p1~p: valid ascii/multibyte chars
+                    // p ~e: invalid bytes + unknown bytes
+                    int clen = enc.maxLength();
+                    if (p1 < p) {
+                        buf = makeConcatNode.executeMake(buf, RopeOperations.create(ArrayUtils.extractRange(pBytes, p1, p), enc, CodeRange.CR_VALID), enc);
```

Comment:
I think instead of `CR_VALID`, we need to pass `CR_UNKNOWN` here and allow a byte scan to occur. It's possible that the substring of `CR_VALID` string is actually `CR_7BIT`.

It's possible this could be modeled as a `SubstringRope` as well and then we wouldn't need to extract anything.